### PR TITLE
HTTP2 settings API

### PR DESCRIPTION
**Pull request by zhaohuabing**
Related: https://github.com/envoyproxy/gateway/issues/3048

### CODE REVIEW

File: api/v1alpha1/clienttrafficpolicy_types.go [hunk 1]
```diff
@@ -200,6 +204,23 @@ type HTTP10Settings struct {
...
+	// +kubebuilder:validation:Maximum=2147483647
+	// +optional
+	InitialStreamWindowSize *uint32 `json:"initialStreamWindowSize,omitempty"`
+
+	// InitialConnectionWindowSize sets the initial window size for HTTP/2 connections.
+	// If not set, the default value is 1 MiB(1024*1024).
+	// +kubebuilder:validation:Minimum=65535
+	// +kubebuilder:validation:Maximum=2147483647
+	// +optional
+	InitialConnectionWindowSize *uint32 `json:"initialConnectionWindowSize,omitempty"`
```
Comment (zhaohuabing) [inline api/v1alpha1/clienttrafficpolicy_types.go:221]: Should EG enforce that connection window size >= stream window size?

File: api/v1alpha1/clienttrafficpolicy_types.go [hunk 2]
```diff
@@ -200,6 +204,23 @@ type HTTP10Settings struct {
 	UseDefaultHost *bool `json:"useDefaultHost,omitempty"`
 }
 
+// HTTP2Settings provides HTTP/2 configuration on the listener.
+type HTTP2Settings struct {
+	// InitialStreamWindowSize sets the initial window size for HTTP/2 streams.
```
Comment (arkodg) [inline api/v1alpha1/clienttrafficpolicy_types.go:209]: can we create a custom type for `resource.Quantiy` defined in https://github.com/envoyproxy/gateway/blob/2c4d385fcfb02841483cc36e0292175f60249bbb/api/v1alpha1/connection_types.go#L24 and reuse it here ?

Comment (zhaohuabing) [reply api/v1alpha1/clienttrafficpolicy_types.go:209]: kubebuilder won't generate validation for alias type. If we define a new type, we need to redefine every `resource.Quantiy` function that EG used for the new type `Size`.


https://github.com/envoyproxy/gateway/actions/runs/8772985239/job/24072530543?pr=3222

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 3]
```diff
@@ -707,6 +707,7 @@ func translateListenerConnection(connection *egv1a1.Connection, httpIR *ir.HTTPL
 
 	if connection.BufferLimit != nil {
 		bufferLimit, ok := connection.BufferLimit.AsInt64()
+		connection.BufferLimit.DeepCopy()
```
Comment (guydc) [inline internal/gatewayapi/clienttrafficpolicy.go:710]: should this be assigned to a var?

File: api/v1alpha1/clienttrafficpolicy_types.go [hunk 4]
```diff
@@ -200,6 +204,19 @@ type HTTP10Settings struct {
...
+	// If not set, the default value is 64 KiB(64*1024).
+	// +optional
+	InitialStreamWindowSize *Size `json:"initialStreamWindowSize,omitempty"`
+
+	// InitialConnectionWindowSize sets the initial window size for HTTP/2 connections.
+	// If not set, the default value is 1 MiB(1024*1024).
+	// +optional
+	InitialConnectionWindowSize *Size `json:"initialConnectionWindowSize,omitempty"`
```
Comment (guydc) [inline api/v1alpha1/clienttrafficpolicy_types.go:217]: Two additional knobs that can be important here are:
- `max_concurrent_streams` - the secure-by-default value of 100 is quite strict, and users may want to increase it.
- `override_stream_error_on_invalid_http_message` - see discussion on how this field should be configured based on wether envoy is a L1 or L2 proxy: https://www.envoyproxy.io/docs/envoy/latest/configuration/best_practices/level_two. In other words, there's no "correct" value here, and we should make it configurable for users based on their topology.

**Review by guydc — COMMENTED** [review]
Thanks! Would love to see two more knobs supported.

File: api/v1alpha1/clienttrafficpolicy_types.go (hunk 4 above)
Comment (zhaohuabing) [reply api/v1alpha1/clienttrafficpolicy_types.go:217]: `override_stream_error_on_invalid_http_message` should be added in http1,http2,and http3 settings. Could we address this in another PR?

Comment (guydc) [reply api/v1alpha1/clienttrafficpolicy_types.go:216]: created #3245

**Review by guydc — APPROVED** [review]
LGTM, Thanks!

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Pull request merged**